CommonItem/src/main/java/com/github/sef24sp4/common/item/loottable/LootTable.java

```diff
@@ -0,0 +1,98 @@
+	 * @param itemChances a map which maps {@link ItemRarity} to their chances. Chances are of type double.
+	 * @throws IllegalArgumentException if the sum of the chances is greater than 1, or any chance is negative or greater than 1
+	 * @see ItemRarity
+	 */
+	public LootTable(Map<ItemRarity, Double> itemChances) {
+		this(itemChances, ServiceLoader.load(ItemSPI.class).stream().map(ServiceLoader.Provider::get).toList());
+	}
+
+	/**
+	 * Constructs a new LootTable with the given item chances and item providers. Only used for testing.
+	 *
+	 * @param itemChances a map from {@link ItemRarity} to their chances of type double.
+	 * @param itemProviders a collection of item providers
+	 * @throws IllegalArgumentException if the sum of the chances is greater than 1, or any chance is negative or greater than 1
+	 * @see ItemRarity
+	 */
+	LootTable(Map<ItemRarity, Double> itemChances, final Collection<ItemSPI> itemProviders) {
+		this.itemProviders = itemProviders;
+		double sumChance = 0;
+		for (Map.Entry<ItemRarity, Double> entry : itemChances.entrySet()) {
+			double chance = entry.getValue();
+			if (chance < 0) throw new IllegalArgumentException();
+			if (chance > 1) throw new IllegalArgumentException();
+			sumChance += chance;
+		}
+		if (sumChance > 1) throw new IllegalArgumentException();
+		this.itemChances = itemChances;
+	}
+
+	/**
+	 * Chooses a rarity based on the item chances. If the sum of the chances is less than 1, the remaining chance is for Optional.empty().
+	 *
+	 * @return an Optional containing the chosen rarity, or an empty Optional if no rarity was chosen
+	 * @see ItemRarity
+	 * @see Optional
+	 */
+	Optional<ItemRarity> chooseRarity() {
+		double random = Math.random();
+		double sum = 0;
+		for (Map.Entry<ItemRarity, Double> entry : this.itemChances.entrySet()) {
+			if (random < entry.getValue() + sum) return Optional.of(entry.getKey());
+			sum += entry.getValue();
+		}
+		return Optional.empty();
+	}
+
+	/**
+	 * Gets an item from the loot table.
+	 * The rarity of the item is chosen using the item chances, and then an item of that rarity is chosen from the item providers.
+	 * If {@link #chooseRarity()} returns an Optional.empty(), this method will also return an Optional.empty().
+	 *
+	 * @return an Optional containing the chosen item, or an empty Optional if no item was chosen.
+	 * @throws Exception if no item providers are available for the chosen rarity.
+	 * @see #chooseRarity()
+	 * @see Optional
+	 */
+	public Optional<CommonItem> getItem() throws Exception {
+		Optional<ItemRarity> chosenRarity = this.chooseRarity();
+		if (chosenRarity.isEmpty()) return Optional.empty();
+		List<ItemSPI> itemSPIList = new ArrayList<>();
+		Map<ItemRarity, Collection<ItemSPI>> provider = generateItemProvidersMap(this.itemProviders);
+		provider.forEach((itemRarity, itemSPIS) -> {
+			if (itemRarity.equals(chosenRarity.get())) itemSPIList.addAll(itemSPIS);
+		});
+		if (itemSPIList.isEmpty()) throw this.emptyItemSPIList;
```

Comment:
Instead of throwning an exception consider returning an empty `Optional` instead.

Reply:
It's more that there should be an SPI if chosenRarity isn't empty

Reply:
The problem is what is the best course of action when such a condition arises? 
Should the game just crash, if there is not a specific item for a given `ItemRarity`?